hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkWriteHelper.java

```diff
@@ -59,10 +59,9 @@ public static SparkWriteHelper newInstance() {
     }).reduceByKey((rec1, rec2) -> {
       @SuppressWarnings("unchecked")
       T reducedData = (T) rec1.getData().preCombine(rec2.getData());
-      // we cannot allow the user to change the key or partitionPath, since that will affect
-      // everything
-      // so pick it from one of the records.
-      return new HoodieRecord<T>(rec1.getKey(), reducedData);
+      HoodieKey reducedKey = rec1.getData().equals(reducedData) ? rec1.getKey() : rec2.getKey();
```

Comment:
so if `preCombine()` is a merge, then the equals check may not succed?

Reply:
Ya I see your point here. From the existing payload implementations `precombine` either returned `this` or `other` so this seemed fine. But if there was other payloads being used with merging behavior, default `HoodieKey` would be set to `rec2.getKey()`. This could be solved by regenerating key here, but there would be performance cost

Reply:
let's roll with this for now. If you don't mind, please raise a JIRA for the follow on.